### ANN Regression

Preprocessing:
1.	Import the libraries and dataset (use tensorflow for Neural Network)
2.	(Optional) Preprocessing steps you think is necessary
3.	Train test split


In [1]:
import numpy as np
import pandas as pd
import tensorflow as tf

df = pd.read_excel('../Data/Folds5x2_pp.xlsx')
df

AT      V       AP     RH      PE
0     14.96  41.76  1024.07  73.17  463.26
1     25.18  62.96  1020.04  59.08  444.37
2      5.11  39.40  1012.16  92.14  488.56
3     20.86  57.32  1010.24  76.64  446.48
4     10.82  37.50  1009.23  96.62  473.90
...     ...    ...      ...    ...     ...
9563  16.65  49.69  1014.01  91.00  460.03
9564  13.19  39.18  1023.67  66.78  469.62
9565  31.32  74.33  1012.92  36.48  429.57
9566  24.48  69.45  1013.86  62.39  435.74
9567  21.60  62.52  1017.23  67.87  453.28

[9568 rows x 5 columns]

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

x = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

Building model:
1.	Initialize the ANN
2.	Adding two dense layer (activation function = relu, and units = 6)
3.	Adding a output layer


In [3]:
ann = tf.keras.models.Sequential()
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))
ann.add(tf.keras.layers.Dense(units=1))

Training model:
1.	Compile the ANN (opttimizer = adam, loss = MSE)
2.	Training ANN (batchsize = 32, epochs = 100)
3.	Predict the test set with your ANN

In [4]:
ann.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])

In [5]:
ann.fit(X_train, y_train, batch_size=32, epochs=100, validation_split=0.2)


Epoch 1/100
192/192 ━━━━━━━━━━━━━━━━━━━━ 1s 677us/step - loss: 206587.4531 - mae: 454.1966 - val_loss: 205097.4688 - val_mae: 452.6009
Epoch 2/100
192/192 ━━━━━━━━━━━━━━━━━━━━ 0s 394us/step - loss: 204199.3125 - mae: 451.6130 - val_loss: 199520.1406 - val_mae: 446.4524
Epoch 3/100
192/192 ━━━━━━━━━━━━━━━━━━━━ 0s 368us/step - loss: 196841.0312 - mae: 443.4122 - val_loss: 184587.3438 - val_mae: 429.3212
Epoch 4/100
192/192 ━━━━━━━━━━━━━━━━━━━━ 0s 369us/step - loss: 179005.6094 - mae: 422.6173 - val_loss: 157379.8594 - val_mae: 395.7742
Epoch 5/100
192/192 ━━━━━━━━━━━━━━━━━━━━ 0s 367us/step - loss: 148640.9531 - mae: 384.0784 - val_loss: 120592.0000 - val_mae: 344.1403
Epoch 6/100
192/192 ━━━━━━━━━━━━━━━━━━━━ 0s 358us/step - loss: 110868.1250 - mae: 328.5327 - val_loss: 82639.2656 - val_mae: 278.5553
Epoch 7/100
192/192 ━━━━━━━━━━━━━━━━━━━━ 0s 357us/step - loss: 74691.8281 - mae: 261.9229 - val_loss: 52426.8047 - val_mae: 210.7483
Epoch 8/100
192/192 ━━━━━━━━━━━━━━━━━━━━ 0s 362us/step - l

In [6]:
y_pred = ann.predict(X_test)
y_pred

60/60 ━━━━━━━━━━━━━━━━━━━━ 0s 503us/step


array([[431.23312],
       [459.24896],
       [463.40076],
       ...,
       [470.38843],
       [442.93182],
       [462.62195]], dtype=float32)

Evaluation:
1.	Report the RMSE and R2 with your model
2.	If your results are not as expected, don’t worry too much. Instead, provide a brief report discussing potential factors that may have contributed to the outcomes.
3.	If your results are good, simply write a two sentence summary

In [8]:
from sklearn.metrics import mean_squared_error, r2_score

rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print("Root Mean Squared Error (RMSE):", rmse)
r2 = r2_score(y_test, y_pred)
print("R-Squared Score:", r2)

Root Mean Squared Error (RMSE): 4.498766513763142
R-Squared Score: 0.9308043012777079


With an RMSE of 4.50 and an R-Squared of 0.93, the model performed well and indicated high accuracy in prediction. The model seems to effectively capture the underlying patterns in the data to reliably predict PE.